src/common/mailing.ts

```diff
@@ -171,3 +179,134 @@ export const addPrivateSourceJoinParams = ({
 
   return urlObj.toString();
 };
+
+const ITEMS_PER_DESTROY = 4000;
+const ITEMS_PER_IDENTIFY = 250;
+
+interface SyncSubscriptionsWithActiveStateProps {
+  con: DataSource;
+  users: GetUsersActiveState;
+}
+
+export const syncSubscriptionsWithActiveState = async ({
+  con,
+  users: { inactiveUsers, downgradeUsers, reactivateUsers },
+}: SyncSubscriptionsWithActiveStateProps) => {
+  await blockingBatchRunner({
+    data: reactivateUsers,
+    runner: async (batch) => {
+      const validReactivateUsers = await con.getRepository(User).find({
+        select: ['id'],
+        where: { id: In(batch), cioRegistered: false },
+      });
+
+      if (validReactivateUsers.length === 0) {
+        return true;
+      }
+
+      await blockingBatchRunner({
+        batchLimit: ITEMS_PER_IDENTIFY,
+        data: validReactivateUsers.map(({ id }) => id),
+        runner: async (ids) => {
+          const users = await con
```

Comment:
Why do we fetch again the users? we fetched them earlier

Reply:
The other fetch is from BQ data. This fetch is internally querying the users from API DB. This saves us time whether we should really process the user or not. For example, this cron failed midway while some users were already processed, fetching here would allow us to avoid running those users again.

Reply:
I agree with this notion but still there's no need to query two times.
Here at first: https://github.com/dailydotdev/daily-api/pull/2550/files/3e718033b199746e7a98eead96c2ab640e7fc0cc#diff-a4ce5390febde3cbc254fdf2d25dec35b0144e25fa3df809578cca6d638a7037R198
And now again in the line of the comment.

Also please select only the fields you need throughout the PR. User object is a big one

Reply:
Ah, yes, this must be a mistake, we don't need this extra fetch.

As for the columns, from the looks of how we pass the values to CIO, we send the whole user object. Unless we explicitly define all the columns to receive by CIO, we can't specify it here as well.